articles/budougumi0617.re

```diff
@@ -1,18 +1,282 @@
+//list[clash][無限ループで完了しないコード]{
+package main
+
+import "fmt"
+
+func goroutineA(c1 chan int, c2 chan int) {
+    _ = <-c2
+    c1 <- 1
+    return
+}
+
+func goroutineB(c1 chan int, c2 chan int) {
+    fmt.Println("c1", <-c1)
+    c2 <- 2
+    return
+}
+
+func main() {
+    c1 := make(chan int)
+    c2 := make(chan int)
+
+    go goroutineA(c1,c2)
+    go goroutineB(c1,c2)
+
+    for {
+    }
+}
+//}
+
+//cmd{
+$ docker run --rm --name sandbox -it golang:1.12.9-buster /bin/bash
+root@d5d112c8c215:/go/src# go get golang.org/x/debug/cmd/viewcore
+root@d5d112c8c215:/go/src# go build -gcflags "-N -l" main.go
+root@d5d112c8c215:/go/src# GOTRACEBACK=crash ./main &
+[1] 786
+root@d5d112c8c215:/go/src# kill -ABRT 786
+root@d5d112c8c215:/go/src# SIGABRT: abort
+PC=0x48739a m=0 sigcode=0
+
+goroutine 1 [running]:
+main.main()
+	/go/src/main.go:22 +0xba fp=0xc000074f98 sp=0xc000074f58 pc=0x48739a
+...
+
+[1]+  Aborted                 (core dumped) GOTRACEBACK=crash ./main
+root@d5d112c8c215:/go/src# viewcore core goroutines
+G stacksize=800
+  000000c000009918 000000c000009938 runtime.usleep+61
+  000000c000009938 000000c0000099b8 runtime.sighandler+976
+  000000c0000099b8 000000c000009a68 runtime.sigtrampgo+498
+  000000c00002c758 000000c00002c798 main.main+190
+  000000c00002c798 000000c00002c7e0 runtime.main+524
+...
+//}
+
+
+コアファイルの取得方法はここ。
+https://golang.org/doc/gdb
+
+=== @<code>{golang.org/x/exp}パッケージ
+https://godoc.org/golang.org/x/exp
+https://github.com/golang/exp
+
+試験的に実装されたコードが内包されています。
+例を挙げると、x/xerrorsも元はこのリポジトリで実装されていました。
+定期的にチェックしておくと、Goの新しい機能を先取りできるかも？
+
+```bash 
+$ git log --oneline wrap.go
+3ee3066 xerrors: remove useless condition checking in Is
+3850056 xerrors: fix Is panics if target is uncomparable
+1f06c39 xerrors: fix crash in As
+a5947ff xerrors: require that error be type assertable to As's target
+334af84 xerrors: check for nil pointers in As's target
+4e09f5a xerrors: make As match on assignability
+20feca1 xerrors: copied files from x/exp/errors
+```
+
+=== @<code>{golang.org/x/image}パッケージ
+ * https://godoc.org/golang.org/x/image
+ * https://github.com/golang/image
+
+標準パッケージでサポートしていない形式の実装も入っている。
+フォントもサポートしていたり、drawサブパッケージが便利。
+https://qiita.com/tebakane/items/b7a47379659d42364c8d
+https://text.baldanders.info/golang/resize-image/
+
+=== @<code>{golang.org/x/lint}パッケージ
+ * https://godoc.org/golang.org/x/lint
+ * https://github.com/golang/lint
+
+golintコマンドの実装
+
+=== @<code>{golang.org/x/mobile}パッケージ
+ * https://godoc.org/golang.org/x/mobile
+ * https://github.com/golang/mobile
+
+gomobileの実装
+GoでiOSアプリやAndroidアプリを作ることができる
+
+=== @<code>{golang.org/x/mod}パッケージ
+ * https://godoc.org/golang.org/x/mod
+ * https://github.com/golang/mod
+
+go modulesに関するコードだけど今も最新仕様に追従できているのかは謎。
+ちゃんと更新履歴を確認しておく必要がある。
+
+=== @<code>{golang.org/x/net}パッケージ
+ * https://godoc.org/golang.org/x/net
+ * https://github.com/golang/net
+
+ネットワーク関連のexpな実装が多く含まれている。http2、websocketとか。 
+ちょっとサンプルコード書きたいなー。
+もしくはgRPCとかkubernetsで使われているかな？
```

Comment:
**[textlint]** <sub>reported by [reviewdog](https://github.com/reviewdog/reviewdog) :dog:</sub>
Disallow to use "？". (ja-technical-writing/no-exclamation-question-mark)